caput/memh5.py

```diff
@@ -2503,13 +2762,135 @@ def _copy_to_file(memgroup, h5group):
-    comm=None,
+    mode,
     hints=True,
     convert_dataset_strings=False,
     convert_attribute_strings=True,
-    **kwargs
+    compression=None,
+    compression_opts=None,
+    **_,
+):
+    """Private routine to copy full data tree from distributed memh5 object into a Zarr file.
+
+    This paralellizes all IO."""
+
+    if not zarr_available:
+        raise RuntimeError("Can't write to zarr file. Please install zarr.")
+
+    # == Create some internal functions for doing the read ==
+    # Function to perform a recursive clone of the tree structure
+    def _copy_to_file(memgroup, group, compression, compression_opts):
+
+        # Copy over attributes
+        if memgroup.comm.rank == 0:
+            copyattrs(
+                memgroup.attrs, group.attrs, convert_strings=convert_attribute_strings
+            )
+
+        # Sort the items to ensure we insert in a consistent order across ranks
+        for key in sorted(memgroup):
+
+            item = memgroup[key]
+
+            # If group, create the entry and the recurse into it
+            if is_group(item):
+                if memgroup.comm.rank == 0:
+                    group.create_group(key)
+                memgroup.comm.Barrier()
+                _copy_to_file(item, group[key], compression, compression_opts)
+
+            # If dataset, create dataset
+            else:
+                if item.compression is not None:
+                    compression = item.compression
+                    compression_opts = item.compression_opts
+
+                # Check if we are in a distributed dataset
+                if isinstance(item, MemDatasetDistributed):
+
+                    data = check_unicode(item)
+
+                    # Write to file from MPIArray
+                    data.to_file(
+                        group,
+                        key,
+                        chunks=item.chunks,
+                        compression=compression,
+                        compression_opts=compression_opts,
+                        file_format=fileformats.Zarr,
+                    )
+                    dset = group[key]
+
+                    if memgroup.comm.rank == 0 and hints:
+                        dset.attrs["__memh5_distributed_dset"] = True
+
+                # Create common dataset (collective)
+                else:
+
+                    # Convert from unicode to bytestring
+                    if convert_dataset_strings:
+                        data = ensure_bytestring(item.data)
+                    else:
+                        data = check_unicode(item)
+
+                    # Write common data from rank 0
+                    if memgroup.comm.rank == 0:
+                        dset = group.create_dataset(
+                            key,
+                            shape=data.shape,
+                            dtype=data.dtype,
+                            chunks=item.chunks,
+                            **fileformats.Zarr.compression_kwargs(
```

Comment:
I would say, we should ensure every `create_dataset` has the `fileformats.compression_kwargs` treatment...